### PR TITLE
Refactor/dp3053 performance improvements to the broadcast notification

### PR DESCRIPTION
**Pull request by youwithouto**

### CODE REVIEW

File: notification/error.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		AndroidNotificationResponseReasonInvalidRegistration: NotificationRequestErrorReasonAndroidInvalidRegistration,
+		AndroidNotificationResponseReasonInternalServerError: NotificationRequestErrorReasonAndroidInternalServerError,
+	}
+
+	ErrInvalidNotificationErrorReason = errors.New("invalid Notification Error Reason")
+	ErrInvalidNotificationDeviceOS    = errors.New("invalid Notification Device OS")
+)
+
+func (e DataNotificationError) Error() string {
+	return fmt.Sprintf("failed sending data notification to %v tokens", len(e.NotificationErrors))
```
Comment (jakequade-hsm) [inline notification/error.go:73]: worth us giving them the error reason?

File: notification/token.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		ErrorMsg  *string  `json:"error_msg"`
+		UpdatedAt int64    `json:"updated_at"`
+	}
+)
+
+const (
+	IOS     = MobileOS("ios")
+	Android = MobileOS("android")
+
+	createTokenTable = `CREATE TABLE IF NOT EXISTS notification_token
```
Comment (jakequade-hsm) [inline notification/token.go:27]: what's the thinking re creating tables programmatically? Do we not have migrations that solve this for us?

File: notification/token.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+// findTokens returns the tokens of userID
+// or all tokens if passing nil
```
Comment (jakequade-hsm) [inline notification/token.go:115]: do we want all tokens to be accessible at once?

**Review by jakequade-hsm — APPROVED** [review]
minor comments, lgtm

File: notification/error.go (hunk 1 above)
Comment (youwithouto) [reply notification/error.go:73]: This info is backend only. The entire API call is async and will always be successful. This error is used to communicate the error state between different backend services.

File: notification/token.go (hunk 2 above)
Comment (youwithouto) [reply notification/token.go:27]: My thinking is that creating tables programmatically is combining table creation with service deployment, i.e., no need to manually creating the tables beforehand.

Whenever we need to update a table, we have two options: a) use the migration hook; b) do a manual update. 

For option a), it requires us to define all steps programmatically (which is not always possible), and there will be a "downtime" when the updated service is being deployed, and this hook will always be called whenever the service is redeployed. However, this does not involve a manual step. 

For option b), it is a manual step but always works, and can minimise "downtime".

I always use option b) for DB table updates.

File: notification/token.go (hunk 3 above)
Comment (youwithouto) [reply notification/token.go:115]: I guess so, as it's part of the requirement for sending the notifications.

The RESTful API allows us to send push notifications to one, a selected group, or all members.

**Pull request merged**